lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+        end_date = Date.parse(raw_reservation[:end_date])
+
+        range = DateRange.new(start_date, end_date)
+
+        parsed_reservation = {
+          id: raw_reservation[:id].to_i,
+          block_id: block_id,
+          room: room,
+          date_range: range
+        }
+
+        reservation = Reservation.new(parsed_reservation)
+        reservations << reservation
+      end
+      return reservations
+    end
+
+    def find_block_id(id)
+      check_id(id)
+      return @blocks.find { |block| block.id == id }
+    end
+
+    def find_room(room_num)
+      check_room_num(room_num)
+      return @rooms.find { |room| room.room_num == room_num }
+    end
+
+    def list_reservations_by_date(specified_date)
+      reservations_by_date = @reservations.find_all do |reservation|
+        date_range = reservation.date_range.get_range
+        date_range.include?(specified_date)
+      end
+      return reservations_by_date
+    end
+
+    def reservations_overlaps?(requested_dates)
+      matching_reservations = @reservations.find_all do |reservation|
+        date_range = reservation.date_range
+        reservation if date_range.overlaps?(requested_dates)
+      end
+    end
+
+    def find_reserved_rooms(requested_dates)
+      matching_reservations = reservations_overlaps?(requested_dates)
+
+      reserved_rooms = matching_reservations.map do |reservation|
+        reservation.room
+      end
+      return reserved_rooms
+    end
+
+    def find_blocks(requested_dates)
+      blocks_by_date = @blocks.map do |room|
+        date_range = room.date_range
+        room if date_range.overlaps?(requested_dates)
+      end
+      return blocks_by_date.compact
+    end
+
+    def find_unavailable_rooms(requested_dates)
+      reserved_rooms = find_reserved_rooms(requested_dates)
+      blocks = find_blocks(requested_dates)
+      unavailable_rooms = reserved_rooms + blocks
+      return unavailable_rooms
+    end
+
+    def find_available_rooms(requested_dates)
+      unavailable_rooms = find_unavailable_rooms(requested_dates)
+      available_rooms = @rooms.reject { |room| unavailable_rooms.include?(room) }
+      check_availablity?(available_rooms)
+      return available_rooms
+    end
+
+    def get_first_available_room(requested_dates)
+      all_available_rooms = find_available_rooms(requested_dates)
+      first_available_room = all_available_rooms.first
+      return first_available_room
+    end
+
+    def confirm_valid_dates?(start_date, end_date)
+      check_dates_validity?(start_date, end_date)
+      check_dates_order?(start_date, end_date)
+    end
+
+    def get_requested_dates(start_date, end_date)
+      check_prior_today?(start_date)
+      confirm_valid_dates?(start_date, end_date)
+      return DateRange.new(start_date, end_date)
+    end
+
+    def reserve_room(input)
+      requested_dates = get_requested_dates(input[:start_date], input[:end_date])
+
+      reservation_id = @reservations.length + 1
```

Comment:
I like the idea of using hash initialization here. However, modern Ruby style is to use keyword arguments to solve the same problem. See https://robots.thoughtbot.com/ruby-2-keyword-arguments#keyword-arguments-vs-options-hash